src/creatureevent.cpp

```diff
@@ -29,15 +29,21 @@ CreatureEvents::CreatureEvents() :
 	scriptInterface.initState();
 }
 
-void CreatureEvents::clear()
+void CreatureEvents::clear(bool fromLua)
 {
 	//clear creature events
-	for (auto& it : creatureEvents) {
-		it.second.clearEvent();
+	for (auto it = creatureEvents.begin(); it != creatureEvents.end(); ) {
+		if (fromLua == it->second.isFromLua()) {
+			it = creatureEvents.erase(it);
+		} else {
+			++it;
+		}
 	}
 
 	//clear lua state
-	scriptInterface.reInitState();
+	if (!fromLua) {
+		scriptInterface.reInitState();
+	}
```

Comment:
Hm, this seems to be very repetitive. I suggest you move it to the parent and call it on derived classes. Also, create a function for the `for` above instead of repeating it everywhere.

Reply:
this part can be moved that's ok but for the `for` loops they're actually not all the same so it wouldn't make much sense to move them to elswhere, as they already have their function which is explicit for them.